### PR TITLE
Sprayer: split Enable from on/off and add simulator

### PR DESCRIPTION
**Pull request by peterbarker**
Before this series of commits, it is not possible to change the sprayer parameters without enabling the output of the sprayer.  This constitutes not just a bad interface, but a safety and possible moral hazard.

The cause of this is the sharing of the "ENABLE"  flag for both "parameter group enable" and "distribute tasty, tasty Brawndo to surrounding plants".

This series of commits splits "ENABLE" into two concepts; parameters should be shown (and sprayer subsystem enabled) and "run", which permits the sprayer to move.

In addition to that split, a simulated sprayer is introduced to help test the sprayer code.  The SITL-simulated multicopter now has a sprayer attachment; you can improve the simulated vehicle's performance by distributing electrolytes to surrounding (simulated) plants.

Please merge; it's what drones crave.

### CODE REVIEW

**Comment by magicrub** [comment]
I don't see a description so I guess we're supposed to ignore this PR? :)

**Comment by peterbarker** [comment]
@magicrub Description added.  If no description==no-merging then long description==rapid-merging? :-)

File: libraries/SITL/SIM_Sprayer.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    void update(const struct Aircraft::sitl_input &input);
+
+    float payload_mass() const { return capacity; }; // kg; assume water
+
+private:
+
+    const uint32_t report_interval = 1000000; // microseconds
+    uint64_t last_report_us;
+
+    // const float pump_max_rate = 0.1; // litres/second
```
Comment (OXINARF) [inline libraries/SITL/SIM_Sprayer.h]: Leftover?

File: libraries/SITL/SIM_Sprayer.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    Sprayer(const uint8_t _pump_servo, int8_t _spinner_servo) :
+        pump_servo(_pump_servo),
+        spinner_servo(_spinner_servo)
+    {}
+
+    // update sprayer state
+    void update(const struct Aircraft::sitl_input &input);
+
+    float payload_mass() const { return capacity; }; // kg; assume water
```
Comment (OXINARF) [inline libraries/SITL/SIM_Sprayer.h]: kg or litres?

File: libraries/SITL/SIM_Sprayer.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    // const float pump_max_rate = 0.1; // litres/second
+    const float pump_max_rate = 0.01; // litres/second
+    const float pump_slew_rate = 20; // percent/scond
+    float last_pump_output; // percentage
+
+    const float spinner_max_rate = 3600; // degrees/second
+    const float spinner_slew_rate = 20; // percent/second
+    float last_spinner_output; // percentage
+
+    // double capacity = 10; // litres
```
Comment (OXINARF) [inline libraries/SITL/SIM_Sprayer.h]: Leftover?

File: libraries/AC_Sprayer/AC_Sprayer.cpp [hunk 4]
```diff
@@ -66,34 +66,43 @@ AC_Sprayer::AC_Sprayer(const AP_InertialNav* inav) :
...
 /// update - adjust pwm of servo controlling pump speed according to the desired quantity and our horizontal speed
 void
 AC_Sprayer::update()
 {
-    // exit immediately if we are disabled (perhaps set pwm values back to defaults)
-    if (!_enabled) {
+    // exit immediately if we are disabled or shouldn't be running
+    if (!_enabled || !running()) {
+        if (running()) {
+            run(false);
```
Comment (OXINARF) [inline libraries/AC_Sprayer/AC_Sprayer.cpp]: The if can be removed, the *run* correctly checks if it is already not running and does nothing in that case.

**Comment by magicrub** [comment]
>If no description==no-merging then long description==rapid-merging?

It's more like a divide by zero problem :)

File: libraries/SITL/SIM_Sprayer.h (hunk 2 above)
Comment (peterbarker) [inline libraries/SITL/SIM_Sprayer.h]: kilograms; the "assume water" indicates that it is 1:1 kg:mass

File: libraries/AC_Sprayer/AC_Sprayer.cpp (hunk 4 above)
Comment (peterbarker) [inline libraries/AC_Sprayer/AC_Sprayer.cpp]: Fair call.

**Comment by peterbarker** [comment]
@OXINARF Fixes pushed up for the issues you raised - thanks!

File: libraries/SITL/SIM_Sprayer.h (hunk 1 above)
Comment (peterbarker) [inline libraries/SITL/SIM_Sprayer.h]: Yeah.  For debugging purposes it is useful to have more in the tank and higher flow rate.

Worked well on the virtual ground, but it was probably asking a little much for the little multicopter to take off with 10kg.  I'll remove the commented lines.

File: libraries/SITL/SIM_Sprayer.h (hunk 2 above)
Comment (peterbarker) [inline libraries/SITL/SIM_Sprayer.h]: I've tried to make the comment clearer.

**Comment by OXINARF** [comment]
@peterbarker Besides the *fixup!* commit that needs to be squashed I say this is ready to go in.

**Comment by rmackay9** [comment]
Merged, thanks!

**Pull request closed without merging**